src/main/scala/com/fulcrumgenomics/util/DemuxFastqs.scala

```diff
@@ -0,0 +1,538 @@
+  */
+package com.fulcrumgenomics.util
+
+import java.io.File
+import java.nio.file.Path
+
+import com.fulcrumgenomics.cmdline.{FGBioTool, ClpGroups}
+import com.fulcrumgenomics.util.miseq.{Sample, SampleBarcode, SampleSheet}
+import dagr.commons.CommonsDef.{DirPath, FilePath, PathToFastq, _}
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.fastq.{FastqReader, FastqRecord}
+import htsjdk.samtools.metrics.MetricsFile
+import htsjdk.samtools.util._
+import picard.illumina.ExtractIlluminaBarcodes
+import picard.illumina.parser.ReadStructure
+import picard.util.IlluminaUtil
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+object DemuxFastqs {
+  val UnmatchedSampleId: String = "unmatched"
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** Creates the read name to put into a SAM record from a FASTQ record. */
+  def extractName(header: String): String = header.takeWhile(ch => ch != ' ').stripSuffix("/1").stripSuffix("/2")
+
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call. Observed
+    * qualities may not be null. */
+  private[util] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
+    observedBases.zip(expectedBases).count {
+      case (observedBase, expectedBase) => !SequenceUtil.isNoCall(expectedBase) && !SequenceUtil.basesEqual(observedBase, expectedBase)
+    }
+  }
+
+  /** Gets all the bases (concatenated) for the given cycle indexes (1-based), ignoring cycles not after the end
+    * of `bases`.  The cycles must be in ascending order.
+    */
+  private def getBasesAtCycles(bases: Array[Byte], cycles: Array[Int]): Array[Byte] = {
+    //  ignore cycles past the end of `bases`.
+    cycles.filter(_ <= bases.length).map(cycle => bases(cycle-1))
+  }
+
+  /** Sets the sample barcode in the given sample sheet using the given read structures and bases from the sample sheet.
+    *
+    * The final sample barcode is the concatenation of the sample barcode from the i7 and i5 bases found in the sample
+    * sheet in that order.  The concatenation uses `IlluminaUtil.barcodeSeqsToString`.  The specific bases extracted
+    * from the i7 and i5 reads are the sample barcode bases specified in the given i7 and i5 read structures.  The
+    * final sample barcode is then updated in the given sample sheet.  An exception is thrown if no sample barcode
+    * bases are found in the i5 or i7 read structures.
+    */
+  private def setSampleBarcode(sampleSheet: SampleSheet, r1ReadStructure: ReadStructure, r2ReadStructure: ReadStructure, i7ReadStructure: ReadStructure, i5ReadStructure: ReadStructure) {
```

Comment:
This one might be worth fixing sooner rather than later for two reasons:

1. This could be done as case classes and make it really nice to use in match statements etc.
2. I would really like to extend it to have syntax for `*` or `_` or something to indicate _all_ or _all remaining_ bases, e.g. `8B*T`